homeassistant/components/mqtt/__init__.py

```diff
@@ -724,53 +610,101 @@ class Subscription:
         self._paho_lock = asyncio.Lock()
 
-        if protocol == PROTOCOL_31:
+        self.init_client()
+        self.config_entry.add_update_listener(self.async_config_entry_updated)
+
+    @staticmethod
+    async def async_config_entry_updated(hass, entry) -> None:
+        """Handle signals of config entry being updated.
+
+        This is a static method because a class method (bound method), can not be used with weak references.
+        Causes for this is config entry options changing.
+        """
+        self = hass.data[DATA_MQTT]
+
+        conf = hass.data.get(DATA_MQTT_CONFIG)
+        if conf is None:
+            conf = CONFIG_SCHEMA({DOMAIN: dict(entry.data)})[DOMAIN]
+
+        self.conf = _merge_config(entry, conf)
+        await self.async_disconnect()
+        self.init_client()
+        await self.async_connect()
+
+        await discovery.async_stop(hass)
+        if self.conf.get(CONF_DISCOVERY):
+            await _async_setup_discovery(hass, self.conf, entry)
+
+    def init_client(self):
+        """Initialize paho client."""
+        # We don't import on the top because some integrations
+        # should be able to optionally rely on MQTT.
+        import paho.mqtt.client as mqtt  # pylint: disable=import-outside-toplevel
+
+        if self.conf[CONF_PROTOCOL] == PROTOCOL_31:
             proto: int = mqtt.MQTTv31
         else:
             proto = mqtt.MQTTv311
 
+        client_id = self.conf.get(CONF_CLIENT_ID)
         if client_id is None:
             self._mqttc = mqtt.Client(protocol=proto)
         else:
             self._mqttc = mqtt.Client(client_id, protocol=proto)
 
+        username = self.conf.get(CONF_USERNAME)
+        password = self.conf.get(CONF_PASSWORD)
         if username is not None:
             self._mqttc.username_pw_set(username, password)
 
+        certificate = self.conf.get(CONF_CERTIFICATE)
+
+        # For cloudmqtt.com, secured connection, auto fill in certificate
+        if (
+            certificate is None
+            and 19999 < self.conf[CONF_PORT] < 30000
+            and self.conf[CONF_BROKER].endswith(".cloudmqtt.com")
+        ):
+            certificate = os.path.join(
+                os.path.dirname(__file__), "addtrustexternalcaroot.crt"
+            )
+
+        # When the certificate is set to auto, use bundled certs from requests
+        elif certificate == "auto":
+            certificate = requests.certs.where()
+
+        # Be able to override versions other than TLSv1.0 under Python3.6
+        conf_tls_version: str = self.conf.get(CONF_TLS_VERSION)
+        if conf_tls_version == "1.2":
+            tls_version = ssl.PROTOCOL_TLSv1_2
+        elif conf_tls_version == "1.1":
+            tls_version = ssl.PROTOCOL_TLSv1_1
+        elif conf_tls_version == "1.0":
+            tls_version = ssl.PROTOCOL_TLSv1
+        else:
+            # Python3.6 supports automatic negotiation of highest TLS version
+            if sys.hexversion >= 0x03060000:
```

Comment:
We can remove this check, we're always on Py3.6+

Reply:
Makes me even wonder if we just drop all the other options, as we can now figure it out ?

Reply:
Sure, I'll do that in a separate PR though since it's a breaking change.